### ISSUE TITLE
Extraneous move in DefaultIterableDiffer (or non optimal ngFor)

### ISSUE THREAD

**Issue by staeke** [angular/angular#7689]
Type: bug (or possibly non-optimal behavior)
Plunkr: http://plnkr.co/edit/WHJs6F4lfkDZNyrQHYbQ?p=preview
Version (UPDATED): Angular 2 beta 11

The current combined behavior for ```DefaultIterableDiffer```/```ngFor``` currently exposes this bad behavior:
1. remove the head item in an array
2. On check, the ```DefaultIterableDiffer``` will now produce a "removed" record for the first item and "moved" records for the rest
3. This leads ```ngFor``` to remove all items, and then add back the non-removed items, thereby leading to a lot of unnecessary DOM removals and adds, losing DOM state (such as focused elements/selection etc.) ( see https://github.com/angular/angular/blob/f9fb72fb0e9bcbda7aeebbf8321ce5d70d78ecee/modules/angular2/src/common/directives/ng_for.ts#L106)

On a related note, the current diff structure doesn't impose any necessary order between removals, moves, and adds (each having an independent forEach method). (There's just a todo by rado on improving this.) Which means that it's unclear (aside from any documentation notes) in which order they should be applied for indices to be valid. To me it seems much more logical that the ```DefaultIterableDiffer``` deduces an **array** of operations to perform on the previous data structure to reach the current state. If you accept that notion, the moves are extraneous and can be removed. Which is something I'd like to propose. Thus, a head removal would just translate to something like:

```
[{type: 'removed', item: 'A', previousIndex=0, ... }]
```

...rather than (the currently implicit through different forEachXXX)
```
{
  removed: [{type: 'removed', item: 'A', previousIndex=0, ... }],
  moved: [
    {type: 'moved', item: 'B', previousIndex=1, currentIndex=0, ... },
    {type: 'moved', item: 'C', previousIndex=2, currentIndex=1, ... },
  ]
}
```

**Comment by joeyrobert** [comment]
Wouldn't this cause huge performance issues with `*ngFor`, particularly with large lists? We're noticing this too at Rangle with iframes causing unnecessary re-renders (as in #8011). Could someone with experience in Angular 2's array diffing algorithm chime in? It would be much appreciated.

I've updated the Plunker to Angular 2 RC -- the issue remains: http://plnkr.co/edit/REflr2JZpDo0a7zSwBUr?p=preview

**Comment by joeyrobert** [comment]
As a follow up to this, adding trackBy appears to show the incorrect behavior as well:

http://plnkr.co/edit/HtZn5HjdnUA9teeNLa8t?p=preview

Typing in the second input "B" removes "A" but then moves focus to "C" when it should still be on "B" (because that's what it's tracking).

**Comment by zoechi** [comment]
I think this was fixed by #10287

**Comment by vicb** [comment]
thanks @zoechi
